plugins/parsers/avro/parser.go

```diff
@@ -0,0 +1,238 @@
+package avro
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/parsers"
+	"github.com/jeremywohl/flatten"
+	"github.com/linkedin/goavro/v2"
+	"time"
+)
+
+type Parser struct {
+	MetricName      string   `toml:"metric_name"`
+	SchemaRegistry  string   `toml:"avro_schema_registry"`
+	Schema          string   `toml:"avro_schema"`
+	Measurement     string   `toml:"avro_measurement"`
+	Tags            []string `toml:"avro_tags"`
+	Fields          []string `toml:"avro_fields"`
+	Timestamp       string   `toml:"avro_timestamp"`
+	TimestampFormat string   `toml:"avro_timestamp_format"`
+	DiscardArrays   bool     `toml:"avro_discard_arrays"`
+	FieldSeparator  string   `toml:"avro_field_separator"`
+	DefaultTags     map[string]string
+	TimeFunc        func() time.Time
+
+	Log telegraf.Logger `toml:"-"`
+}
+
+func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	var schema string
+
+	if len(buf) < 5 {
+		err := fmt.Errorf("buf is %d bytes; must be at least 5", len(buf))
+		return nil, err
+	}
+	schemaID := int(binary.BigEndian.Uint32(buf[1:5]))
+	binaryData := buf[5:]
```

Comment:
Shoudn't the underlying parser library deal with this?

Reply:
If you're doing the usual use case, where you have a schema registry and the binary data corresponding to the packed encoding of that schema, sure.  But in the parent PR, there was discussion from someone who wanted to manually specify the schema rather than retrieve it from the registry.  Apparently not all Avro users are also Kafka users.

See https://github.com/influxdata/telegraf/pull/7732#discussion_r463325701 .  Maybe @rodrigobrochado is still around and can comment?  Anyway, I put in the guard so we had a slightly more controlled error if somehow we got passed a short set of bytes (rather than taking a slice that wasn't there), which was @rodrigobrochado's suggestion on the original PR.

In the not-Kafka case, I don't actually know if I should just decode the whole object, rather than just the sixth and subsequent bytes, because it's not my use case.  I can change to that if you think it's a good idea.

Reply:
@athornton and @rodrigobrochado thanks for clarifying! Can you please make this a small helper function e.g. `extractScheme(buf []byte) (int, []byte, error)` and explain what you are doing with a small comment?